File: json/src/main/java/top/andnux/json/JsonConverterFactory.java
package top.andnux.json;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

public final class JsonConverterFactory extends Converter.Factory {

    private JsonAdapter mJsonAdapter;

    public JsonConverterFactory(JsonAdapter jsonAdapter) {
        mJsonAdapter = jsonAdapter;
    }

    public static JsonConverterFactory create(JsonAdapter converter) {
        return new JsonConverterFactory(converter);
    }


    public static JsonConverterFactory create() {
        return new JsonConverterFactory(new FastJsonAdapter());
    }

    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(Type type,
                                                            Annotation[] annotations,
                                                            Retrofit retrofit) {
        return new JsonResponseBodyConverter<>(mJsonAdapter,type);
    }

    @Override
    public Converter<?, RequestBody> requestBodyConverter(Type type,
                                                          Annotation[] parameterAnnotations,
                                                          Annotation[] methodAnnotations,
                                                          Retrofit retrofit) {
        return new JsonResponseBodyConverter<>(mJsonAdapter,type);
    }
}

File: net/src/main/java/top/andnux/net/retrofit/adapter/JsonConverterFactory.kt
package top.andnux.net.retrofit.adapter

import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import top.andnux.json.FastJsonProxy
import top.andnux.json.JsonProxy
import java.lang.reflect.Type

class JsonConverterFactory(private val mJsonProxy: JsonProxy) : Converter.Factory() {

    override fun responseBodyConverter(type: Type,
                                       annotations: Array<Annotation>?,
                                       retrofit: Retrofit?): Converter<ResponseBody, *> {
        return JsonResponseBodyConverter<Any>(mJsonProxy, type)
    }

    override fun requestBodyConverter(type: Type,
                                      parameterAnnotations: Array<Annotation>?,
                                      methodAnnotations: Array<Annotation>?,
                                      retrofit: Retrofit?): Converter<*, RequestBody> {
        return JsonResponseBodyConverter(mJsonProxy, type)
    }

    companion object {
        @JvmStatic
        @JvmOverloads
        fun create(jsonProxy: JsonProxy = FastJsonProxy()): JsonConverterFactory {
            return JsonConverterFactory(jsonProxy)
        }
    }
}
